test/python/algorithms/test_grover.py

```diff
@@ -104,19 +104,51 @@ def test_implicit_phase_oracle_is_good_state(self):
         self.assertEqual(result.top_measurement, "0")
 
     def test_fixed_iterations(self):
-        """Test the iterations argument"""
+        """Test the iterations argument with an int"""
         grover = Grover(iterations=2, quantum_instance=self.statevector)
         problem = AmplificationProblem(Statevector.from_label("111"), is_good_state=["111"])
         result = grover.amplify(problem)
         self.assertEqual(result.top_measurement, "111")
 
+    def test_fixed_iterations_without_good_state(self):
+        """Test the iterations argument with an int and without good state"""
+        grover = Grover(iterations=2, quantum_instance=self.statevector)
+        problem = AmplificationProblem(Statevector.from_label("111"))
+        result = grover.amplify(problem)
+        self.assertEqual(result.top_measurement, "111")
+
+    def test_no_iterations_with_good_state(self):
+        """Test with a good state argument and without iterations"""
+        grover = Grover(quantum_instance=self.statevector)
+        problem = AmplificationProblem(Statevector.from_label("111"), is_good_state=["111"])
+        result = grover.amplify(problem)
+        self.assertEqual(result.top_measurement, "111")
+
+    def test_no_iterations_without_good_state(self):
+        """Test that correct error is thrown without iterations and good state argument"""
+        grover = Grover(quantum_instance=self.statevector)
+        problem = AmplificationProblem(Statevector.from_label("111"))
+        with self.assertRaisesRegex(
+            TypeError, "An is_good_state function is required with the provided oracle"
+        ):
```

Comment:
Personally, I usually just put
```python
with self.assertRaises(TypeError):
```
but I don't know if this regex version is better practice -- I'll leave you to decide 😄

Reply:
Yeah in this case just `assertRaises` is probably fine, the regex check probably adds more value if you have multiple possible `TypeError`s or if your error message contains a variable you want to check. But anyway I don't think it hurts to use the stricter checking so I'll leave it here 😄

Reply:
`assertRaisesRegex` is probably the better habit to get into - it's good to check that later changes to the code don't cause an unrelated error to start getting shown to the user without the author _deciding_ to make that change.  It's not a big deal either way, though.